Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -293,6 +295,318 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntry* e
+{
+  TextureConfig config;
+  p.Do(config);
+
+  std::vector<u8> texture_data;
+  p.Do(texture_data);
+
+  if (p.GetMode() != PointerWrap::MODE_READ)
+    return nullptr;
+
+  std::unique_ptr<AbstractTexture> tex = g_texture_cache->AllocateTexture(config);
+  if (!tex)
+  {
+    PanicAlert("Failed to create texture for deserialization");
+    return nullptr;
+  }
+
+  size_t start = 0;
+  for (u32 layer = 0; layer < config.layers; layer++)
+  {
+    for (u32 level = 0; level < config.levels; level++)
+    {
+      u32 level_width = std::max(config.width >> level, 1u);
+      u32 level_height = std::max(config.height >> level, 1u);
+      size_t stride = AbstractTexture::CalculateStrideForFormat(config.format, level_width);
+      size_t size = stride * level_height;
+      if ((start + size) > texture_data.size())
+      {
+        ERROR_LOG(VIDEO, "Insufficient texture data for layer %u level %u", layer, level);
+        return tex;
+      }
+
+      // TODO: Load multiple layers.
+      if (layer != 0)
+        tex->Load(level, level_width, level_height, level_width, &texture_data[start], size);
+      start += size;
+    }
+  }
+
+  return tex;
+}
+
+void TextureCacheBase::DoState(PointerWrap& p)
+{
+  p.Do(frameCount);
+  p.Do(last_entry_id);
+
+  if (p.GetMode() == PointerWrap::MODE_WRITE || p.GetMode() == PointerWrap::MODE_MEASURE)
+    DoSaveState(p);
+  else
+    DoLoadState(p);
+}
+
+void TextureCacheBase::DoSaveState(PointerWrap& p)
+{
+  std::map<TCacheEntry*, u32> entry_map;
+  auto AddCacheEntryToMap = [&entry_map, &p](TCacheEntry* entry) {
+    auto iter = entry_map.find(entry);
+    if (iter != entry_map.end())
+      return;
+
+    // Skip custom textures since they won't fit in our staging texture.
+    if (entry->is_custom_tex)
+      return;
+
+    u32 id = static_cast<u32>(entry_map.size());
+    entry_map.emplace(entry, id);
+  };
+  auto GetCacheEntryId = [&entry_map](TCacheEntry* entry) -> u32 {
+    auto iter = entry_map.find(entry);
+    return iter != entry_map.end() ? iter->second : static_cast<u32>(-1);
+  };
+
+  // Collect all referenced texture cache entries and assign them IDs.
+  for (const auto& it : textures_by_address)
+    AddCacheEntryToMap(it.second);
+  for (const auto& it : textures_by_hash)
+    AddCacheEntryToMap(it.second);
+
+  // Save the texture cache entries out in the order the were referenced.
+  u32 size = static_cast<u32>(entry_map.size());
+  p.Do(size);
+  for (const auto& it : entry_map)
+  {
+    TCacheEntry* entry = it.first;
+    TCacheEntry::DoState(p, &entry);
+  }
+  p.DoMarker("TextureCacheEntries");
+
+  // Save references for each cache entry.
+  // As references are circular, we need to have everything created before linking entries.
+  std::set<std::pair<u32, u32>> reference_pairs;
+  for (const auto& it : entry_map)
+  {
+    TCacheEntry* entry = it.first;
+    u32 id1 = GetCacheEntryId(entry);
+    for (TCacheEntry* referenced_entry : entry->references)
```

Comment:
`referenced_entry` can be a const pointer, assuming `GetCacheEntryId`'s parameter is made const.